OpenAPI support for kubectl

pwittrock opened the pull request:
Support for openapi spec in kubectl.

Includes:
- downloading and caching openapi spec to a local file
- parsing openapi spec into binary serializable datastructures (10x faster load times 600ms -> 40ms)
- caching parsed openapi spec in memory for each command

```release-note
NONE
```

pwittrock committed:
kubectl OpenAPI support

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44531)
<!-- Reviewable:end -->

pwittrock committed:
kubectl OpenAPI bazel updates

pwittrock committed:
kubectl OpenAPI whitelist pkg/kubectl/cmd/util/openapi for linting

pwittrock committed:
kubectl OpenAPI godeps updates

pwittrock commented:
@k8s-bot verify test this

```
W0416 18:58:18.522] fatal: unable to access 'https://bitbucket.org/bertimus9/systemstat.git/': The requested URL returned error: 502
W0416 18:58:18.522] godep: error downloading dep (bitbucket.org/bertimus9/systemstat): exit status 128
W0416 18:58:24.761] godep: Error downloading some deps. Aborting restore and check.
W0416 18:58:24.765] !!! Error in ./hack/godep-restore.sh:28
```

pwittrock commented:
@caesarxuchao - plz review the client-go changes
@mbohlool - plz review the openapi parsing
@fabianofranz / @monopole - plz review for kubectl

mbohlool commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -401,6 +401,11 @@ func AddValidateFlags(cmd *cobra.Command) {
 }
 
+func AddOpenAPIFlags(cmd *cobra.Command) {
+	cmd.Flags().String("schema-cache-dir", fmt.Sprintf("~/%s/%s", clientcmd.RecommendedHomeDir, clientcmd.RecommendedSchemaName), fmt.Sprintf("If non-empty, load/store cached API schemas in this directory, default is '$HOME/%s/%s'", clientcmd.RecommendedHomeDir, clientcmd.RecommendedSchemaName))
```
I've seen three reference to `schema`, `openapi`, and `swagger` in code/comments/messages. should we just use one? I suggest `openapi` for all of them.

pwittrock replied:
This doesn't introduce a new flag, it is just allows registering the cache dir without registering the validation flags.  See 5 lines above this one.

mbohlool commented on pkg/kubectl/cmd/util/openapi/openapi.go:
```diff
@@ -0,0 +1,308 @@
+
+// isPrimitive returns true if s is a primitive type
+func (o *Resources) isPrimitive(s spec.Schema) bool {
+	return !o.isArray(s) && !o.isMap(s) && !o.isDefinitionReference(s) && len(s.Type) > 0
```
The assumption of a schema is primitive if it does not have a `ref` may not be true all the time. I would say a primitive schema is one with no "properties". does that make sense?

pwittrock replied:
Updated the checks to be more explicit.

There is a unit check that test that every definition encountered has exactly 1 of (isPrimitive, isArray, isMap, isDefinitionReference) which should give some confidence that these are working as intended.

mbohlool commented on pkg/kubectl/cmd/util/openapi/openapi_cache.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+// encodeSpec binary encodes the openapi spec
+func (c *cachingOpenAPIClient) encodeSpec(parsed *Resources) ([]byte, error) {
```
Can't we just encode/decode `spec.swagger`?

pwittrock replied:
No.  :( As discussed offline, the Ref type which contains the json reference has private fields that are dropped by binary serialization.

monopole commented:
Reviewed 3 of 3 files at r1, 19 of 19 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[pkg/kubectl/cmd/testing/fake.go, line 406 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxACINac-NPJS1vdYv:-KhxACINac-NPJS1vdYw:bz224h9) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/testing/fake.go#L406)):*
> ```Go
> func (f *FakeFactory) SwaggerSchema(schema.GroupVersionKind) (*swagger.ApiDeclaration, error) {
> 	return nil, nil
> }
> ```

should the swagger stuff be marked deprecated and/or be given a TODO(#issue) to track its removal?

---

*[pkg/kubectl/cmd/util/factory.go, line 445 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxAso-92dK8CthlLl4:-KhxAso-92dK8CthlLl5:b-e3wat1) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/factory.go#L445)):*
> ```Go
> 	cacheFile := path.Join(fullDir, prefix, groupVersion, schemaFileName)
> 
> 	abs, _ := filepath.Abs(cacheDir)
> ```

this can be scoped to the 'if' body

---

*[pkg/kubectl/cmd/util/factory_object_mapping.go, line 442 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxBmmcpNVdIfWV5BxP:-KhxBmmdnvq6GeaPizcW:b83y7i3) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/factory_object_mapping.go#L442)):*
> ```Go
> 
> // openAPIData returns an openAPIData containing metadata and structural information about Kubernetes
> // models and operations.  Will try to cache the spec locally if possible.
> ```

pls add note about what (if anything) will make the cache stale, and how it is freshened

test coverage for that?

---

*[pkg/kubectl/cmd/util/helpers.go, line 405 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxCfmxBocu3DhP2r__:-KhxCfmyciXQqzrx2wYu:b-8zinjc) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/helpers.go#L405)):*
> ```Go
> 
> func AddOpenAPIFlags(cmd *cobra.Command) {
> 	cmd.Flags().String("schema-cache-dir", fmt.Sprintf("~/%s/%s", clientcmd.RecommendedHomeDir, clientcmd.RecommendedSchemaName), fmt.Sprintf("If non-empty, load/store cached API schemas in this directory, default is '$HOME/%s/%s'", clientcmd.RecommendedHomeDir, clientcmd.RecommendedSchemaName))
> ```

BTW , go fmt will retain newlines added for readability. 

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 30 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxG8h4scNq4DpAjrUq:-KhxG8h4scNq4DpAjrUr:b-bzy60r) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L30)):*
> ```Go
> )
> 
> const gvkOpenAPIKey = "x-kubernetes-group-version-kind"
> ```

Please comment the const.  Does the x mean something?  could this string be arbitrary, e.g. peach-banana-sandwich?

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 40 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxI-ISRsYHxEL4r2WE:-KhxI-ISRsYHxEL4r2WF:bjgljd) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L40)):*
> ```Go
> }
> 
> // KindDefinition defines a Kubernetes object Kind
> ```

Seems odd to have PeachDefinition instead of just Peach.  The doc doesn't help.
Maybe for doc inspiration see https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html

Went digging in  apimachinery/blob/master/pkg/runtime/schema/group_version.go and am even more confused about Kind

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 42 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxJjTduzB0R4OinYm1:-KhxG_dreWdLwPaJoLHo:bl1vyba) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L42)):*
> ```Go
> // KindDefinition defines a Kubernetes object Kind
> type KindDefinition struct {
> 	// Name is the name of the type
> ```

BTW, this looks like L65, hence earlier request for commentary on Kind. 
"TypeDefinition" below raises the same questions

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 48 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxGtK9T-HNIcCBE29H:-KhxGtK9T-HNIcCBE29I:b-vdgcbf) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L48)):*
> ```Go
> 	IsResource bool
> 
> 	// GroupVersionKind is the group version kind of a resource.
> ```

// GroupVersionKind (does this useful thing in the context) of a resource.

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 63 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxHrO-Co6DE0oTktuM:-KhxHrO0GhqVJlxgc-D8:b-mruvqx) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L63)):*
> ```Go
> }
> 
> // TypeDefinition defines a field definition
> ```

"defines a Type"

and again wondering why not just "Type" and how this differs from "Kind".

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 193 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxM7Hb7-bgTtUXGz83:-KhxM7Hb7-bgTtUXGz84:b-c7q2jp) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L193)):*
> ```Go
> // isArray returns true if s is an array type.
> func (o *Resources) isArray(s spec.Schema) bool {
> 	return len(s.Type) > 0 && s.Type[0] == "array"
> ```

BTW: Could "ARRAY" show up?

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 196 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxMICpjgctQB3eZry2:-KhxMICpjgctQB3eZry3:b-o73lf5) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L196)):*
> ```Go
> }
> 
> // isMap
> ```

isObject?  (given s.Type[0] == "object")

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 232 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxMdaFdEj8vnQLi0Ts:-KhxMdaFdEj8vnQLi0Tt:bu722xs) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L232)):*
> ```Go
> }
> 
> // getElementType returns the type of an element for arrays and maps
> ```

Comment looks odd given !o.isArray returns error

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 259 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxNBDErwv9uToJhzye:-KhxNBDF8Iji03Tr-k2C:bv34ywp) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L259)):*
> ```Go
> 
> // getGroupVersionKind implements openAPIData
> // getGVK parses the gropuversionkind for a resource definition from the x-kubernetes
> ```

sic (gropuversionkind)

 More and more want a primer in this doc about Group, Kind, Version, Type, Field
(or a ptr to a primer).   

Doc under https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/schema/group_version.go doesn't help
as much as i'd hoped.

---

*[pkg/kubectl/cmd/util/openapi/openapi_cache.go, line 48 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxRRd6UUxTeWrezwP5:-KhxRRd6UUxTeWrezwP6:bmv7ayk) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi_cache.go#L48)):*
> ```Go
> // newCachingOpenAPIClient returns a new discovery.OpenAPISchemaInterface
> // that will read the openapi spec from a local cache if it exists, and
> // if not fetch an openapi spec using client
> ```

using client.


People are not picky about sentence formation?

---

*[pkg/kubectl/cmd/util/openapi/openapi_cache_test.go, line 60 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxXdn_6t2WC4LV7nTK:-KhxXdn_6t2WC4LV7nTL:b-ahszy8) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi_cache_test.go#L60)):*
> ```Go
> 		result, err := instance.openAPIData()
> 		Expect(err).To(BeNil())
> 		expectEqual(result, expectedData)
> ```


FYI: make an Equals method for Resources, test it in openapi_test.go, and use it directly here?  instead of the expectEqual utility below?

---

*[pkg/kubectl/cmd/util/openapi/openapi_test.go, line 67 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxYijX7rTOy37Q8AW2:-KhxYijX7rTOy37Q8AW3:bkx14ru) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi_test.go#L67)):*
> ```Go
> 	})
> 
> 	It("should find the definition fields", func() {
> ```

BTW: nifty framework :+1:   tests easy to read

---

*[staging/src/k8s.io/client-go/discovery/discovery_client.go, line 380 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxZ52qOieZRKZo1Zk3:-KhxZ52qOieZRKZo1Zk4:b-zbh42s) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/staging/src/k8s.io/client-go/discovery/discovery_client.go#L380)):*
> ```Go
> 	// WARNING: We don't use the loads.Analyze because this is very expensive
> 	// and loads.Analyze takes 2x as long (it Unmarshal's 2x)
> 	// Warning: this is very expensive (~600ms)
> ```

1) comment above function?
2) loads.Analyzed called below, seems to contradict comment (though technically it differs by one letter)
3) Warning: twice?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531)*
<!-- Sent from Reviewable.io -->

arschles commented on pkg/kubectl/cmd/util/openapi/doc.go:
```diff
@@ -0,0 +1,20 @@
+package openapi
+
+// The openapi package is a collection of libraries for fetching the openapi spec
+// from a Kubernetes server and then indexing the type definitions.
```
if these are meant to be Godoc-compatible comments, they need to go above the `package` keyword

pwittrock commented:
Review status: 20 of 28 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/kubectl/cmd/testing/fake.go, line 406 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxACINac-NPJS1vdYv:-KhyK-2pETU1kNyAFgLF:b-8b11j9) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/testing/fake.go#L406)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

should the swagger stuff be marked deprecated and/or be given a TODO(#issue) to track its removal?
</blockquote></details>

Added

---

*[pkg/kubectl/cmd/util/factory.go, line 445 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxAso-92dK8CthlLl4:-KhyKL2g-3EFYfc7H6UZ:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/factory.go#L445)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

this can be scoped to the 'if' body
</blockquote></details>

Done

---

*[pkg/kubectl/cmd/util/factory_object_mapping.go, line 442 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxBmmcpNVdIfWV5BxP:-KhyLbkx8OYmJgzw63kp:b75vebt) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/factory_object_mapping.go#L442)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

pls add note about what (if anything) will make the cache stale, and how it is freshened

test coverage for that?


</blockquote></details>

Updated with a comment.

Test coverage for the caching logic is in the openapi_cache_test.go file

---

*[pkg/kubectl/cmd/util/helpers.go, line 405 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxCfmxBocu3DhP2r__:-KhyMB0U7mGC2OejXdPQ:b3uca6b) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/helpers.go#L405)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

BTW , go fmt will retain newlines added for readability. 
</blockquote></details>

Added some newlines for readability

---

*[pkg/kubectl/cmd/util/openapi/doc.go, line 20 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BQ3-r3-20:-KhyMeA-XItA81gIshzu:b-q562yl) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/doc.go#L20)):*
<details><summary><i>Previously, arschles (Aaron Schlesinger) wrote…</i></summary><blockquote>

if these are meant to be Godoc-compatible comments, they need to go above the `package` keyword
</blockquote></details>

Done thx.

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 30 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxG8h4scNq4DpAjrUq:-KhyN0i8HZfhC7vJPzKq:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L30)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

Please comment the const.  Does the x mean something?  could this string be arbitrary, e.g. peach-banana-sandwich?
</blockquote></details>

Done

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 40 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxI-ISRsYHxEL4r2WE:-KhyRXDOubux2FONoO0e:b7oowuj) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L40)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

Seems odd to have PeachDefinition instead of just Peach.  The doc doesn't help.
Maybe for doc inspiration see https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html

Went digging in  apimachinery/blob/master/pkg/runtime/schema/group_version.go and am even more confused about Kind


</blockquote></details>

Updated to drop the Definition piece.  GVK deserves a good doc.

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 42 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxJjTduzB0R4OinYm1:-KhxwUX6T2yL7hFQAddK:b532xng) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L42)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

BTW, this looks like L65, hence earlier request for commentary on Kind. 
"TypeDefinition" below raises the same questions

</blockquote></details>

Yes it is a bit confusing.  Added a comment to clarify

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 48 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxGtK9T-HNIcCBE29H:-Khxw7Monoc6cGH6VD3I:buw47t2) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L48)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

// GroupVersionKind (does this useful thing in the context) of a resource.
</blockquote></details>

Updated the comment

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 63 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxHrO-Co6DE0oTktuM:-KhxuPMZ8pO2zWAv8QCA:b-4bq7qh) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L63)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

"defines a Type"

and again wondering why not just "Type" and how this differs from "Kind".

</blockquote></details>

Updated the comment.  Let me know if this makes more sense.

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 193 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxM7Hb7-bgTtUXGz83:-Khxtq117DRcw4nxeHYC:b39yi3a) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L193)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

BTW: Could "ARRAY" show up?
</blockquote></details>

Changed to compare against lowercase.  I don't think that this should actually show up, but provides a bit of future proofing.

I've changed the matching logic a bit so it should match one of the primitive types "string", "boolean", "integer" or "array", "object", or be empty for definition references.  The unit test parses the k8s swagger.json and validates that each definition matches exactly one of primitive, array, map, definitionReference, so we should catch if there is anything that isn't looking correct.

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 196 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxMICpjgctQB3eZry2:-KhxtDbRlF0Ohn2zuNMf:bz64z93) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L196)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

isObject?  (given s.Type[0] == "object")
</blockquote></details>

 Updated with a comment to explain this.  Also moved the strings into constants with descriptions.

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 232 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxMdaFdEj8vnQLi0Ts:-KhxtBY-uik7MYEuYZvs:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L232)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

Comment looks odd given !o.isArray returns error
</blockquote></details>

Done

---

*[pkg/kubectl/cmd/util/openapi/openapi.go, line 259 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxNBDErwv9uToJhzye:-KhxsKGdQmNQCxV5Eo8k:bu21gb2) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi.go#L259)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

sic (gropuversionkind)

 More and more want a primer in this doc about Group, Kind, Version, Type, Field
(or a ptr to a primer).   

Doc under https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/schema/group_version.go doesn't help
as much as i'd hoped.


</blockquote></details>

GroupVersionKind is a more general API concept.  There is a little bit [here](https://kubernetes.io/docs/concepts/overview/kubernetes-api/).  Could probably deserve something more in depth.

---

*[pkg/kubectl/cmd/util/openapi/openapi_cache.go, line 48 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxRRd6UUxTeWrezwP5:-Khxs5h3X5WkLulNyBpn:bqf7eow) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi_cache.go#L48)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

using client.


People are not picky about sentence formation?

</blockquote></details>

Sometimes they are :)

---

*[pkg/kubectl/cmd/util/openapi/openapi_cache_test.go, line 60 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxXdn_6t2WC4LV7nTK:-KhyVMNULpAyotdl4Oi1:b2nhpa7) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi_cache_test.go#L60)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

FYI: make an Equals method for Resources, test it in openapi_test.go, and use it directly here?  instead of the expectEqual utility below?
</blockquote></details>

Thx

---

*[staging/src/k8s.io/client-go/discovery/discovery_client.go, line 380 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxZ52qOieZRKZo1Zk3:-KhxrTXMkymoc1wGOBDu:b-z9wjs5) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/staging/src/k8s.io/client-go/discovery/discovery_client.go#L380)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

1) comment above function?
2) loads.Analyzed called below, seems to contradict comment (though technically it differs by one letter)
3) Warning: twice?

</blockquote></details>

:P.  Changed the logic after I started caching.  Updated the comments to be correct.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
PTAL

---

Review status: 20 of 28 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/kubectl/cmd/util/openapi/openapi_test.go, line 67 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxYijX7rTOy37Q8AW2:-KhyWDPr4WUWn6W5tL6o:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi_test.go#L67)):*
<details><summary><i>Previously, monopole (Jeff Regan) wrote…</i></summary><blockquote>

BTW: nifty framework :+1:   tests easy to read
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-:-KhyWF2Ug_pkQCknWoLe:bb71pro)*
<!-- Sent from Reviewable.io -->

fabianofranz commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -417,6 +421,7 @@ func writeSchemaFile(schemaData []byte, cacheDir, cacheFile, prefix, groupVersio
 	if _, err := io.Copy(tmpFile, bytes.NewBuffer(schemaData)); err != nil {
 		return err
 	}
+	glog.Infof("Wrinting swagger cache (dir %v) file %v (from %v)", cacheDir, cacheFile, tmpFile.Name())
```
Use a leveled (`.V(n)`) instead. Also typo "Wrinting".

pwittrock replied:
Suggestion on V level?

fabianofranz replied:
We still have to document log levels but for this I'd say 6 or 7.

pwittrock replied:
Fixed

fabianofranz commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -46,13 +48,22 @@ import (
 	clientAccessFactory ClientAccessFactory
+
+	// openAPIGetter loads and caches openapi specs
+	openAPIGetter openAPIGetter
```
@deads2k ring level looks good?

deads2k replied:
> @deads2k ring level looks good?

Yes.  It looks correct.

fabianofranz commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -427,3 +438,41 @@ func (f *ring1Factory) SwaggerSchema(gvk schema.GroupVersionKind) (*swagger.ApiD
+		// Get the server version for caching the openapi spec
+		versionString := ""
+		version, err := discovery.ServerVersion()
+		if version != nil && err != nil {
```
Not sure I understood, do you ever get both version and error here?

pwittrock replied:
Fixed

fabianofranz commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -427,3 +438,41 @@ func (f *ring1Factory) SwaggerSchema(gvk schema.GroupVersionKind) (*swagger.ApiD
+		// Get the cache directory for caching the openapi spec
+		cacheDir, err = substituteUserHome(cacheDir)
+		if err != nil {
+			// Don't cache the result if we couldn't subsistute the home directory
```
typo "subsistute"

pwittrock replied:
Fixed

fabianofranz commented on staging/src/k8s.io/client-go/discovery/discovery_client.go:
```diff
@@ -365,6 +374,21 @@ func (d *DiscoveryClient) SwaggerSchema(version schema.GroupVersion) (*swagger.A
 }
 
+// OpenAPISchema fetches the open api schema using a rest client and parses the json.
+// Warning: this is very expensive (~1.2s)
```
Is this because of `RawMessage`, or the OpenAPI loading? Kind of worrisome depending on where we'll use it (e.g. for `get`). I remember a significant performance degradation after we switched `get` to use unstructured objects, mostly because of the json decoding to `map[string]interface{}`.

pwittrock replied:
This is because the OpenAPI spec is a massive json file that takes a long time to deserialize into binary form.  This will be cached between calls as a binary, which loads much faster (~40ms).

fabianofranz commented on pkg/kubectl/cmd/util/openapi/openapi_cache.go:
```diff
@@ -0,0 +1,193 @@
+	// No cached version found, download from server
+	s, err := c.client.OpenAPISchema()
+	if err != nil {
+		glog.Warningf("Failed to download openapi data %v", err)
```
Same, leveled log or just `fmt` if you want to always print it.

pwittrock replied:
Fixed

fabianofranz commented on pkg/kubectl/cmd/util/openapi/openapi_cache.go:
```diff
@@ -0,0 +1,193 @@
+
+	oa, err := newOpenAPIData(s)
+	if err != nil {
+		glog.Warningf("Failed to parse openapi data %v", err)
```
Same here.

fabianofranz commented on pkg/kubectl/cmd/util/openapi/openapi_cache.go:
```diff
@@ -0,0 +1,193 @@
+		err = c.writeToCache(oa)
+		if err != nil {
+			// Just log an warning, no need to fail the command since we got the data we need
+			glog.Warningf("Unable to cache openapi spec %v", err)
```
Same here.

pwittrock replied:
Fixed

fabianofranz commented on staging/src/k8s.io/client-go/discovery/discovery_client.go:
```diff
@@ -91,6 +94,11 @@ type SwaggerSchemaInterface interface {
 }
 
+type OpenAPISchemaInterface interface {
+	// SwaggerSchema retrieves and parses the swagger API schema the server supports.
```
"OpenAPISchema..."

pwittrock replied:
Fixed

pwittrock commented:
PTAL

fabianofranz commented:
LGTM from a `kubectl` perspective.

mengqiy commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -33,6 +33,7 @@ import (
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
 
+	"github.com/golang/glog"
```
nit: This line should be moved into github import group

mengqiy commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -28,6 +28,8 @@ import (
 
 	"github.com/emicklei/go-restful/swagger"
 
+	"sync"
```
nit: should move this up.

mengqiy reviewed (commented):
Other LGTM

caesarxuchao commented on staging/src/k8s.io/client-go/discovery/discovery_client.go:
```diff
@@ -91,6 +94,11 @@ type SwaggerSchemaInterface interface {
 	SwaggerSchema(version schema.GroupVersion) (*swagger.ApiDeclaration, error)
 }
 
+type OpenAPISchemaInterface interface {
```
nit: for consistency, could you put a comment for the interface?

caesarxuchao commented on staging/src/k8s.io/client-go/discovery/discovery_client.go:
```diff
@@ -365,6 +374,21 @@ func (d *DiscoveryClient) SwaggerSchema(version schema.GroupVersion) (*swagger.A
+// OpenAPISchema fetches the open api schema using a rest client and parses the json.
+// Warning: this is very expensive (~1.2s)
+func (d *DiscoveryClient) OpenAPISchema() (*spec.Swagger, error) {
+	data, err := d.restClient.Get().AbsPath("/swagger.json").Do().Raw()
```
this is kind of off-topic. /swagger.json is not shown as an path when i visit the root of the server. It only shows "/swaggerapi/" which leads me to swagger-1.2. @mbohlool do you think we should fix that?

caesarxuchao commented on staging/src/k8s.io/client-go/discovery/discovery_client_test.go:
```diff
@@ -325,6 +327,48 @@ func TestGetSwaggerSchemaFail(t *testing.T) {
+}
+
+func TestGetOpenAPISchema(t *testing.T) {
+	expect := spec.Swagger{}
```
Can we add some contents to the spec?

caesarxuchao reviewed (commented):
A couple of nits for the changes in staging/client-go.

deads2k commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -46,13 +48,22 @@ import (
+	openAPIGetter openAPIGetter
+}
+
+type openAPIGetter struct {
```
please avoid the anonymous inclusion in the struct.  It makes the code below harder to read.

deads2k commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -427,3 +438,41 @@ func (f *ring1Factory) SwaggerSchema(gvk schema.GroupVersionKind) (*swagger.ApiD
+	})
+
+	// Delegate to the OpenAPIGetter
+	return f.openAPIGetter.OpenAPIData()
```
This method is on the `Getter` I guess?

deads2k commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -427,3 +438,41 @@ func (f *ring1Factory) SwaggerSchema(gvk schema.GroupVersionKind) (*swagger.ApiD
+		}
+
+		// Create the caching OpenAPIGetter
+		f.openAPIGetter.Getter = openapi.NewOpenAPIGetter(cacheDir, versionString, discovery)
```
Why does it care about the server version?  Why doesn't it use the discovery method and rely on the caching of the discovery client?

pwittrock replied:
Does the discovery client cache the result to disk?

pwittrock replied:
Does the discovery client expose a method to that will give the client stable a hash / key for for exposed API?

monopole commented:
Review status: 20 of 28 files reviewed at latest revision, 24 unresolved discussions.

---

*[pkg/kubectl/cmd/util/openapi/openapi_cache.go, line 48 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhxRRd6UUxTeWrezwP5:-KiGJmgLmNjm-7XuLxB4:bd8ie0t) ([raw file](https://github.com/kubernetes/kubernetes/blob/18d9b0bf58b1a458488480d4fa397e60e11500fe/pkg/kubectl/cmd/util/openapi/openapi_cache.go#L48)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary><blockquote>

Sometimes they are :)
</blockquote></details>

OK :)  it all /lgtm

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531)*
<!-- Sent from Reviewable.io -->

monopole commented:
Reviewed 4 of 8 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
Review status: 22 of 28 files reviewed at latest revision, 23 unresolved discussions.

---

*[pkg/kubectl/cmd/util/factory.go, line 36 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BPx-r5-36:-KiHTj2yI85zbmXMc2f0:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9e315b4f7f77e78f69d8eaa832eec089bb7bb95c/pkg/kubectl/cmd/util/factory.go#L36)):*
<details><summary><i>Previously, mengqiy (Mengqi Yu) wrote…</i></summary><blockquote>

nit: This line should be moved into github import group
</blockquote></details>

Done.

---

*[pkg/kubectl/cmd/util/factory_object_mapping.go, line 64 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BPy-r4-64:-KiHTlgX4Q8kq3VOsCHg:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/be6563675695ac4b5f6628fe7332e69cf9a10017/pkg/kubectl/cmd/util/factory_object_mapping.go#L64)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary><blockquote>

please avoid the anonymous inclusion in the struct.  It makes the code below harder to read.
</blockquote></details>

Done.

---

*[pkg/kubectl/cmd/util/factory_object_mapping.go, line 477 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BPy-r4-477:-KiHTmgF2EFnLJ3wxkuv:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/be6563675695ac4b5f6628fe7332e69cf9a10017/pkg/kubectl/cmd/util/factory_object_mapping.go#L477)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary><blockquote>

This method is on the `Getter` I guess?
</blockquote></details>

Done.

---

*[pkg/kubectl/cmd/util/factory_object_mapping.go, line 31 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BPy-r5-31:-KiHTjvvCCsGCJtGswhv:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9e315b4f7f77e78f69d8eaa832eec089bb7bb95c/pkg/kubectl/cmd/util/factory_object_mapping.go#L31)):*
<details><summary><i>Previously, mengqiy (Mengqi Yu) wrote…</i></summary><blockquote>

nit: should move this up.
</blockquote></details>

Done.

---

*[pkg/kubectl/cmd/util/openapi/openapi_cache.go, line 82 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BQ5-r4-82:-KiHU3Dc2IfRR4mkjlA4:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/be6563675695ac4b5f6628fe7332e69cf9a10017/pkg/kubectl/cmd/util/openapi/openapi_cache.go#L82)):*
<details><summary><i>Previously, fabianofranz (Fabiano Franz) wrote…</i></summary><blockquote>

Same here.
</blockquote></details>

Done.

---

*[staging/src/k8s.io/client-go/discovery/discovery_client.go, line 97 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BQ--r5-97:-KiHToydJGxhX4GILHHC:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9e315b4f7f77e78f69d8eaa832eec089bb7bb95c/staging/src/k8s.io/client-go/discovery/discovery_client.go#L97)):*
<details><summary><i>Previously, caesarxuchao (Chao Xu) wrote…</i></summary><blockquote>

nit: for consistency, could you put a comment for the interface?
</blockquote></details>

Done.

---

*[staging/src/k8s.io/client-go/discovery/discovery_client.go, line 380 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BQ--r5-380:-KiHU5SRSOzgeMdST9yb:b2nha9w) ([raw file](https://github.com/kubernetes/kubernetes/blob/9e315b4f7f77e78f69d8eaa832eec089bb7bb95c/staging/src/k8s.io/client-go/discovery/discovery_client.go#L380)):*
<details><summary><i>Previously, caesarxuchao (Chao Xu) wrote…</i></summary><blockquote>

This comment is off-topic. /swagger.json is not shown as an path when i visit the root of the server. It only shows "/swaggerapi/" which leads me to swagger-1.2. @mbohlool do you think we should fix that?
</blockquote></details>

ACK

---

*[staging/src/k8s.io/client-go/discovery/discovery_client_test.go, line 354 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531#-KhnjU5mqhM4kSbx1BQ0-r5-354:-KiHTqvtAz_AAN1EheFC:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9e315b4f7f77e78f69d8eaa832eec089bb7bb95c/staging/src/k8s.io/client-go/discovery/discovery_client_test.go#L354)):*
<details><summary><i>Previously, caesarxuchao (Chao Xu) wrote…</i></summary><blockquote>

Can we add some contents to the spec?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44531)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
PTAL

pwittrock commented:
@k8s-bot gce etcd3 e2e test this

caesarxuchao commented:
client-go changes lgtm.

pwittrock commented:
@caesarxuchao @deads2k Can I get an approval for client-go?

deads2k commented:
/approve

monopole commented:
/lgtm
